rfcs/20190904-tfx-generic-container-based-component.md

```diff
@@ -0,0 +1,863 @@
+
+    args = parser.parse_args()
+    arguments = args.__dict__
+
+    x_train, y_train, x_eval, y_eval = load_datset(arguments['input_examples'])
+
+    model = LogisticRegression()
+    model.fit(x_train, y_train, args.train_steps)
+
+    #
+    # ... perform grid search, etc, ...
+    #
+
+    # Write the trained model.
+    joblib.dump(model, argument['output_model'])
+```
+
+Given the two (containerized) applications above, our goal is to chain them into
+a 2-step pipeline which will automatically track metadata related to:
+
+*   **Artifacts**: These include the input CSV file, the output CSV file with
+    training and eval examples, and the final Model file.
+*   **Executions**: These include execution steps `MyExampleGen`, with runtime
+    execution property `num_csv_columns`, and `MyTrainer`, with runtime
+    execution property `train_steps`.
+
+Notably, since `my_example_gen.py` produces *Examples*, we make the pipeline
+understand that it is *Examples*, and record it as such in ML Metadata. Doing so
+will enable downstream components, such as Trainer, to understand that it is
+receiving *Examples*, and also realize higher-level functionality such as
+ExampleValidation. Similarly, since `my_trainer.py` produces Model, other parts
+of the pipeline should understand it to enable higher-level functionality such
+as ModelValidation and Pusher to serving.
+
+Hence, we propose to provide an extension to the DSL that allows users to
+declare inputs and outputs in terms of ML Metadata of artifacts during pipeline
+construction. The proposed DSL extension would translate the input and output
+artifacts at pipeline runtime, abstracting reading and writing artifact metadata
+from and to storage on behalf of the wrapped containerized application program.
+
+## Detailed Design
+
+We propose the following syntax for wrapping user’s containers in the DSL by
+introducing `ExecutorSpec`. This syntax complements and extends the
+[ComponentSpec](https://github.com/tensorflow/tfx/blob/4cafb7a3ddd73463d70072dade71f380e86072e6/tfx/types/component_spec.py#L73)
+design previously implemented in TFX, and generalizes
+[EXECUTOR_CLASS](https://github.com/tensorflow/tfx/blob/767968262233bb8d37837e8de9e11d1065b07c7b/tfx/components/base/base_component.py#L66)
+attribute of components. We propose to use the same `ComponentSpec` base class
+to describe the custom container-based component’s input and output artifacts,
+and its parameters.
+
+*   Base `ExecutorSpec` class for container based component.
+
+```python
+class ExecutorSpec(with_metaclass(abc.ABCMeta, object)):
+  """Base class that specifies 'what' to execute for the component."""
+  pass
+
+
+class ExecutorClassSpec(ExecutorSpec):
+  """Execution spec for a Python class-based Executor."""
+
+  # An in-process Python Executor class, derived from TFX
+  # base_executor.Executor class.
+  executor_class = ...
+
+  def __init__(self, executor_class):
+    ...
+
+
+class ExecutorContainerSpec(ExecutorSpec):
```

Comment:
How does this class help a Kubernetes/Kubeflow/KFP user define Kubernetes resources using idiomatic python?

Here's a concrete example. Here's a K8s job to run one of our notebooks to train an xgboost model
https://github.com/kubeflow/examples/blob/master/xgboost_synthetic/testing/job.yaml

So imagine if we wanted to drop that job into a pipeline.

Its a 51 line YAML file. The image and command just cover 2 of those 51 lines.

So to specify the rest of the spec; init containers, environment variables, volume mounts etc...
I assume I would end up just having to put a raw yaml/json text string into the platform_config.

So the DSL/Client library isn't really helping me construct K8s objects by writing idiomatic python.

Furthermore, I'm guessing all those Kubernetes features (init containers, volumes, etc...) would break the ability to run anywhere but Kubernetes (and probably even the K8s cluster where those secrets are installed).

So TFX's support for multiple runners isn't really helping me. And in the meantime, since TFX doesn't want to treat K8s resources as first class entitites because it wants to be runner agnostic. I'm not sure how its ultimately going to be able to help me write idiomatic python code to orchestrate K8s resources.

Reply:
The goal of TFX is not to enable one to write Python code to orchestrate k8s resources. It's to author a logical machine-learning pipeline that is portable. In order to achieve this goal, we need to separate the specification for a logical pipeline from the runtime configuration. When running TFX on Kubeflow Pipelines, one way to achieve this is to supply that configuration to `KubeflowDagRunner`, which is what currently compiles the TFX logical pipeline to the Argo spec that can be deployed on KFP. That separation may look something like this (made up syntax):

```
pipeline = .... # construct logical pipeline
platform_config = {
  'MyTrainer': { 'use_gpu': True, 'memory': '1GB' },   
  ....
}
KubeflowDagRunner.run(pipeline, platform_config=platform_config)  # produce YAML spec
```